MobilityData/mobility-feed-api
# fix: historical Dataset Batch Processing DB Update

**Summary:**

Closes #169 
Part of #105

Sorry if I included too many changes at once; it was hard to have them independently.
Summary
- [x] Moved batch functions inside functions-python. The reasoning is to have a single place for all functions to reuse common patterns for build, testing, formatting, and sharing packages.
- [x] Adding a DB exclusively for testing. As part of each function testing, now it's possible to have more control over what the DB contains, making the tests less likely to fail due to local changes in the DB.
- [x] Created a _database_ utils file for creating DB session and shared across functions. Introduced the concept of reusing the DB connection. This helps to minimize the number of connections as the connections can be reused between different executions of the same function when GCP reuses the same container. The connection should be closed when GCP discards the container, but I haven't tried it yet...
- [x] Reduced the number of GH parameters by moving function-related parameters to JSON files and setting multiple parameters with default values.
- [x] Create a Python class for  the datastore entities
- [x] Reduce the number of HTTP retries. This was causing some feeds to be infinite get processed as the retries exceeded the pub/sub retry timeout.
- [x]  Create a DB URL per environment GCP secret to inject in the Python functions.
- [x] Batch Functions in dev and QA will only process 10 feeds to reduce cost
- [x] Fix: lint script was not failing the build
- [x] Fix: test  script was not failing the build
- [x] Feeds exposed by legacy SSL versions can be downloaded


**Expected behavior:** 

- [x] Feeds are not executed infinite
- [x] Datasets are not published when the previous dataset hash is the same
- [x] Remove the previous dependency on the existence of the dataset from the populate script. 

**Testing tips:**

- Select a feed and change the latest dataset hash field to any string
- On dev, go to google scheduler console 
- Select the `dataset-batch-job-dev` job 
- Click _Resume_
- Click _Force Run_
- Wait for all datasets to be process
- Check _gtfsdataset_ table and verify that a new dataset is created with a valid hash

Please make sure these boxes are checked before submitting your pull request - thanks!
